File: manual/code/generate_strategies/Direct/java/function_patsy_dmatrices.java
package com.example;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.Gson;

public class PatsyProcessor {
    public String process_formula(String data, String formula) {
        // 所有功能均写在此唯一方法中
        try {
            Gson gson = new Gson();
            // 解析JSON字符串到JsonObject
            JsonObject jsonObj = JsonParser.parseString(data).getAsJsonObject();
            // 检查必须存在"y"列且必须为数组
            if (!jsonObj.has("y") || !jsonObj.get("y").isJsonArray()) {
                return "Error: invalid input";
            }
            JsonArray yArray = jsonObj.get("y").getAsJsonArray();
            int nRows = yArray.size();

            // 分割公式，左侧必须为"y"
            String[] parts = formula.split("~");
            if (parts.length != 2) {
                return "Error: invalid input";
            }
            String lhs = parts[0].trim();
            if (!lhs.equals("y")) {
                return "Error: invalid input";
            }
            String rhs = parts[1].trim();

            int xCols = 0;
            // 根据不同的公式模式进行判断并验证必需的列是否存在
            if (rhs.equals("x1 + x2")) {
                if (!jsonObj.has("x1") || !jsonObj.get("x1").isJsonArray() ||
                        !jsonObj.has("x2") || !jsonObj.get("x2").isJsonArray()) {
                    return "Error: invalid input";
                }
                // 截距项 + x1 + x2
                xCols = 3;
            } else if (rhs.equals("x1 * x2")) {
                if (!jsonObj.has("x1") || !jsonObj.get("x1").isJsonArray() ||
                        !jsonObj.has("x2") || !jsonObj.get("x2").isJsonArray()) {
                    return "Error: invalid input";
                }
                // 截距项 + x1 + x2 + 交互项
                xCols = 4;
            } else if (rhs.equals("x1 + I(x1 ** 2) + I(x1 ** 3)")) {
                if (!jsonObj.has("x1") || !jsonObj.get("x1").isJsonArray()) {
                    return "Error: invalid input";
                }
                // 截距项 + x1 + (x1)^2 + (x1)^3
                xCols = 4;
            } else {
                // 公式不匹配的情况返回错误
                return "Error: invalid input";
            }

            // 构造返回结果JSON
            JsonObject result = new JsonObject();
            JsonArray yShape = new JsonArray();
            yShape.add(nRows);
            yShape.add(1);
            JsonArray xShape = new JsonArray();
            xShape.add(nRows);
            xShape.add(xCols);
            result.add("y_shape", yShape);
            result.add("X_shape", xShape);
            return gson.toJson(result);
        } catch (Exception e) {
            return "Error: invalid input";
        }
    }
}

File: code/generate_strategies/IR(CoT)/cpp/cpp/function_patsy_dmatrices.cpp
#include <Eigen/Dense>
#include <nlohmann/json.hpp>
#include <string>
#include <sstream>
#include <stdexcept>
#include <vector>
#include <algorithm>
#include <cmath>

using json = nlohmann::json;

class PatsyProcessor {
public:
    std::string process_formula(const std::string& data, const std::string& formula) {
        try {
            json df = json::parse(data);

            if (!df.contains("y") || !df["y"].is_array()) {
                return "Error: invalid input";
            }

            size_t n = df["y"].size();

            size_t pos = formula.find("~");
            if (pos == std::string::npos) {
                return "Error: invalid input";
            }
            std::string lhs = formula.substr(0, pos);
            std::string rhs = formula.substr(pos + 1);

            auto trim = [](std::string s) -> std::string {
                s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) { return !std::isspace(ch); }));
                s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) { return !std::isspace(ch); }).base(), s.end());
                return s;
            };
            lhs = trim(lhs);
            rhs = trim(rhs);

            if (lhs != "y") {
                return "Error: invalid input";
            }

            Eigen::MatrixXd X;
            int X_cols = 0;

            if (rhs == "x1 + x2") {
                if (!df.contains("x1") || !df.contains("x2")) {
                    return "Error: invalid input";
                }
                X_cols = 3;
                X = Eigen::MatrixXd::Ones(n, X_cols); 
                for (size_t i = 0; i < n; ++i) {
                    X(i, 1) = df["x1"][i].get<double>();
                    X(i, 2) = df["x2"][i].get<double>();
                }
            } else if (rhs == "x1 * x2") {
                if (!df.contains("x1") || !df.contains("x2")) {
                    return "Error: invalid input";
                }
                X_cols = 4;
                X = Eigen::MatrixXd::Ones(n, X_cols);
                for (size_t i = 0; i < n; ++i) {
                    X(i, 1) = df["x1"][i].get<double>();
                    X(i, 2) = df["x2"][i].get<double>();
                    X(i, 3) = df["x1"][i].get<double>() * df["x2"][i].get<double>(); 
                }
            } else if (rhs == "x1 + I(x1 ** 2) + I(x1 ** 3)") {
                if (!df.contains("x1")) {
                    return "Error: invalid input";
                }
                X_cols = 4;
                X = Eigen::MatrixXd::Ones(n, X_cols);
                for (size_t i = 0; i < n; ++i) {
                    double x1_val = df["x1"][i].get<double>();
                    X(i, 1) = x1_val;
                    X(i, 2) = std::pow(x1_val, 2);
                    X(i, 3) = std::pow(x1_val, 3);
                }
            } else {
                return "Error: invalid input";
            }

            json result;
            result["y_shape"] = { static_cast<int>(n), 1 };
            result["X_shape"] = { static_cast<int>(n), X_cols };

            return result.dump();
        } catch (...) {
            return "Error: invalid input";
        }
    }
};